### ISSUE TITLE
High RAM usage in the analysis of COSMX data 

### ISSUE THREAD

**Issue by joaolsf** [HiDiHlabs/sainsc#25]
Hi, I have a COSMX 1K-panel dataset with around 26M transcripts in total. I have run this dataset just fine through FICTURE analysis in my local machine (M1 Max chip, 64GB RAM). I am testing SAINSC and I can see peaking at almost 500GB of RAM usage during calculate_total_mRNA() or calculate_total_mRNA_KDE(), then ofc the kernel breaks. I will run it on a HPC, but I wanted to ask if there is something wrong I am doing, or if there is something that could be done to reduce RAM usage during those calculations, as one of the premises of SAINSC is to be scalable and less resource hungry than other tools, such as FICTURE. 

A second question is related to plotting cell types, what is the basis of this plot, I mean, is it plotting at pixel level, transcript level, bin size (500nm) level or at the bandwidth of the gaussian kernel, in pixels or um? In your example, are those dots in the stereoseq cell type plots at the 4um level (bandwidth) or bin size (500nm) level?

Thank you and great tool!

**Comment by naveedishaque** [comment]
Thanks for the interest! 
Based on transcript and gene count, I can't think of an obvious reason for your high memory usage. We certainly want to get to the bottom of this...
Someone more qualified than me will soon consider your problem, but perhaps in the meantime, can you please post the exact commands you are using (ideally all code from loading the data, to the point where you see 500 GB RAM usage) and some more details about your data (e.g., size of tissue section, predicted transcripts per cell)?

**Comment by niklasmueboe** [comment]
Hi,

let's first tackle the second question as this is a bit easier to answer.
The celltype map is assigned and "plotted" at the pixel-level i.e. bin-size (usually 500 nm but this can also be modified by you). This does not mean that this is rendered on your computer at this resolution though. This depends on your matplotlib settings etc. But that is the theoretical resolution you get if you zoom in enough.

Regarding your first question; this does seem to be odd. We definitely have seen much lower memory usage with more transcripts. In this case it would be helpful to see the exact code you are using to see if I can spot any obvious issues. It also does depend a bit on the size of your sample (in terms of pixels; you can get this by just printing your LazyKDE or GridCounts object). But with this large memory usage I would assume that sth else is going on. Can you maybe also provide the versio of Sainsc you are using?

**Comment by joaolsf** [comment]
Hi,

See answers below:


  1.  exact commands you are using (ideally all code from loading the data, to the point where you see 500 GB RAM usage) and some more details about your data (e.g., size of tissue section, predicted transcripts per cell)?

For the codes, I am following the exact codes described in the ReadTheDocs page.
Predicted transcripts per cell ~ 100

import polars as pl
cosmx_file = Path("/data") / "BMN100424_tx_filtered_transcripts.csv"
n_threads = 16
transcripts = (
    pl.read_csv(
        cosmx_file,
        columns=["target", "x_global_px", "y_global_px"],
        schema_overrides={"target": pl.Categorical},
        n_threads=n_threads,
    )
    .rename({"target": "gene", "x_global_px": "x", "y_global_px": "y"})
    .filter(~pl.col("gene").cast(pl.Utf8).str.contains("(BLANK|NegControl)"))
)

from sainsc import GridCounts
from sainsc import LazyKDE
cosmx = LazyKDE.from_dataframe(transcripts, resolution=1000, binsize=0.5, n_threads=n_threads)
cosmx.calculate_total_mRNA() >>> # peaks at 500GB RAM usage here
cosmx.gaussian_kernel(4, unit="um")
cosmx.calculate_total_mRNA_KDE() >>> # peaks at almost 260GB of RAM...


  1.  It also does depend a bit on the size of your sample (in terms of pixels; you can get this by just printing your LazyKDE or GridCounts object).
LazyKDE (16 threads)
    genes: 1000
    shape: (203295, 294620)
    resolution: 500.0 nm / px


  1.  Can you maybe also provide the versio of Sainsc you are using?
I am using sainsc v0.2.1.

Could the issue be because there are 3 tissue sections across a large area? Should I subset the transcript by each  tissue section first or crop the area in the LazyKDE object?

Thank you!
Joao

From: niklasmueboe ***@***.***>
Date: Wednesday, 18 December 2024 at 6:56 pm
To: HiDiHlabs/sainsc ***@***.***>
Cc: Dr Joao Luiz ***@***.***>, Author ***@***.***>
Subject: Re: [HiDiHlabs/sainsc] High RAM usage in the analysis of COSMX data (Issue #25)

Hi,

let's first tackle the second question as this is a bit easier to answer.
The celltype map is assigned and "plotted" at the pixel-level i.e. bin-size (usually 500 nm but this can also be modified by you). This does not mean that this is rendered on your computer at this resolution though. This depends on your matplotlib settings etc. But that is the theoretical resolution you get if you zoom in enough.

Regarding your first question; this does seem to be odd. We definitely have seen much lower memory usage with more transcripts. In this case it would be helpful to see the exact code you are using to see if I can spot any obvious issues. It also does depend a bit on the size of your sample (in terms of pixels; you can get this by just printing your LazyKDE or GridCounts object). But with this large memory usage I would assume that sth else is going on. Can you maybe also provide the versio of Sainsc you are using?

**Comment by niklasmueboe** [comment]
> Could the issue be because there are 3 tissue sections across a large area? Should I subset the transcript by each  tissue section first or crop the area in the LazyKDE object?

Yes, the issue is that the total area is enormous but there is probably a lot of empty space in between. I'd recommend splitting it into the 3 sections and process them separately.

As a further comment; regarding the loading of your data you are following example code that was written for Xenium. You adjusted it mostly but this specific line

`.filter(~pl.col("gene").cast(pl.Utf8).str.contains("(BLANK|NegControl)"))`

is also technology dependent. These are the names/prefixes that Xenium is using for their control probes. You might need to adjust them to filter out the control probes in COSMX.

**Comment by joaolsf** [comment]
Hi,

The transcript file I’m loading is already filtered so that specific line of code is probably not doing anything, hence I didn’t adjust.

Ok, I’ll try with splitting the file across the 3 tissue sections and process them separately. You are right, there’ll be lots of empty space, still I didn’t have this issue when running Ficture. It ran in the exact big space just fine, and quickly, in my local machine.

Thank you!
Joao

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by niklasmueboe** [comment]
So to the best of my knowledge, Ficture should basically only scale with the number of transcripts. Sainsc on the other hand does depend o the analysed area as well, as you have just witnessed. So yes in the case where most of the sample is empty space ficture will be much more efficient.

But just as a reference, with the size of the sample you had the RGB image of the celltype map at full resolution would already be ~200GB

**Comment by joaolsf** [comment]
Hi,

Thanks for the quick support and I’ll test again with the transcript files by sample.

From: niklasmueboe ***@***.***>
Date: Wednesday, 18 December 2024 at 8:27 pm
To: HiDiHlabs/sainsc ***@***.***>
Cc: Dr Joao Luiz ***@***.***>, Author ***@***.***>
Subject: Re: [HiDiHlabs/sainsc] High RAM usage in the analysis of COSMX data (Issue #25)

So to the best of my knowledge, Ficture should basically only scale with the number of transcripts. Sainsc on the other hand does depend on the analysed area as well, as you have just witnessed. So yes in the case where most of the sample is empty space ficture will be much more efficient.

But just as a reference, with the size of the sample you had the RGB image of the celltype map at full resolution would already be ~200GB
